consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -604,10 +605,36 @@ func (x *XDPoS_v2) voteHandler(chain consensus.ChainReader, voteMsg *utils.Vote)
 	Function that will be called by votePool when it reached threshold.
 	In the engine v2, we will need to generate and process QC
 */
-func (x *XDPoS_v2) onVotePoolThresholdReached(chain consensus.ChainReader, pooledVotes map[common.Hash]utils.PoolObj, currentVoteMsg utils.PoolObj) error {
+func (x *XDPoS_v2) onVotePoolThresholdReached(chain consensus.ChainReader, pooledVotes map[common.Hash]utils.PoolObj, currentVoteMsg utils.PoolObj, proposedBlockHeader *types.Header) error {
+
+	masternodes := x.GetMasternodes(chain, proposedBlockHeader)
+
+	// Filter out non-Master nodes signatures
+	var wg sync.WaitGroup
+	wg.Add(len(pooledVotes))
+	validSignatureMap := map[common.Hash]utils.Signature{}
```

Comment:
question: is this `validSignatureMap` safe when many routines use it?

Reply:
if not, I think use `chan` is better

Reply:
yes, it is safe if there is no risk on writing in same time. Also chan is hard to use on this case, chan is better if we know the number of item will put into the `chan`, but with current implementation, len of validSignatures is dynamic.

Reply:
but I think it may write at same time, this line:
`validSignatureMap[hash] = v.Signature`

Reply:
is this hash means block hash or vote hash?

Reply:
if it's block hash, then we need to atomic this operation. If it's vote, then every vote from different nodes should have different hash

Reply:
ok, I googled about `map` and it says it's not safe...